File: firmware/esp32-s3-cam/src/INA219.cpp
#include "INA219.h"

#define REG_CONFIG        0x00
#define REG_SHUNT_VOLTAGE 0x01
#define REG_BUS_VOLTAGE   0x02
#define REG_POWER         0x03
#define REG_CURRENT       0x04
#define REG_CALIBRATION   0x05

INA219::INA219(uint8_t address, float shuntOhms)
  : _wire(nullptr),
    _address(address),
    _shuntOhms(shuntOhms),
    _currentLsb(0.0001f),
    _powerLsb(_currentLsb * 20.0f) {}

bool INA219::begin(TwoWire& wire) {
  _wire = &wire;

  _wire->beginTransmission(_address);
  if (_wire->endTransmission() != 0) {
    return false;
  }

  calibrate();

  // 32V range, gain /8, 12-bit bus+shunt ADC, continuous mode
  write16(REG_CONFIG, 0x399F);

  return true;
}

void INA219::calibrate() {
  uint16_t cal = (uint16_t)(0.04096f / (_currentLsb * _shuntOhms));
  write16(REG_CALIBRATION, cal);
}

void INA219::write16(uint8_t reg, uint16_t value) {
  _wire->beginTransmission(_address);
  _wire->write(reg);
  _wire->write((value >> 8) & 0xFF);
  _wire->write(value & 0xFF);
  _wire->endTransmission();
}

uint16_t INA219::read16(uint8_t reg) {
  _wire->beginTransmission(_address);
  _wire->write(reg);
  _wire->endTransmission(false);

  _wire->requestFrom((int)_address, 2);
  uint16_t value = ((uint16_t)_wire->read() << 8) | _wire->read();
  return value;
}

int16_t INA219::read16Signed(uint8_t reg) {
  return (int16_t)read16(reg);
}

INA219Reading INA219::read() {
  calibrate();

  uint16_t rawBus = read16(REG_BUS_VOLTAGE);
  int16_t rawShunt = read16Signed(REG_SHUNT_VOLTAGE);
  int16_t rawCurrent = read16Signed(REG_CURRENT);
  uint16_t rawPower = read16(REG_POWER);

  INA219Reading r;
  r.busVoltageV = ((rawBus >> 3) * 0.004f);
  r.shuntVoltageMv = rawShunt * 0.01f;
  r.currentMa = rawCurrent * _currentLsb * 1000.0f;
  r.powerMw = rawPower * _powerLsb * 1000.0f;
  return r;
}

File: esp32-s3-cam/archive/MicroPython/ina219.py
# ina219.py
# Minimal MicroPython INA219 reader

import time

_REG_CONFIG = 0x00
_REG_SHUNT_VOLTAGE = 0x01
_REG_BUS_VOLTAGE = 0x02
_REG_POWER = 0x03
_REG_CURRENT = 0x04
_REG_CALIBRATION = 0x05

class INA219:
    def __init__(self, i2c, address=0x40, shunt_ohms=0.1):
        self.i2c = i2c
        self.address = address
        self.shunt_ohms = shunt_ohms

        # Simple default calibration:
        # current_lsb = 0.1 mA/bit
        self.current_lsb = 0.0001
        self.power_lsb = self.current_lsb * 20

        cal = int(0.04096 / (self.current_lsb * self.shunt_ohms))
        self._write16(_REG_CALIBRATION, cal)

        # 32V range, gain /8, 12-bit bus+shunt ADC, continuous mode
        self._write16(_REG_CONFIG, 0x399F)

    def _write16(self, reg, value):
        self.i2c.writeto_mem(
            self.address,
            reg,
            bytes([(value >> 8) & 0xFF, value & 0xFF])
        )

    def _read16(self, reg, signed=False):
        data = self.i2c.readfrom_mem(self.address, reg, 2)
        value = (data[0] << 8) | data[1]
        if signed and value & 0x8000:
            value -= 65536
        return value

    def bus_voltage_v(self):
        raw = self._read16(_REG_BUS_VOLTAGE)
        return ((raw >> 3) * 0.004)

    def shunt_voltage_mv(self):
        raw = self._read16(_REG_SHUNT_VOLTAGE, signed=True)
        return raw * 0.01

    def current_ma(self):
        # Re-write calibration before reading current; INA219 can reset it.
        cal = int(0.04096 / (self.current_lsb * self.shunt_ohms))
        self._write16(_REG_CALIBRATION, cal)
        raw = self._read16(_REG_CURRENT, signed=True)
        return raw * self.current_lsb * 1000

    def power_mw(self):
        raw = self._read16(_REG_POWER)
        return raw * self.power_lsb * 1000

    def read(self):
        return {
            "kind": "ina219",
            "address": self.address,
            "bus_voltage_v": round(self.bus_voltage_v(), 3),
            "shunt_voltage_mv": round(self.shunt_voltage_mv(), 3),
            "current_ma": round(self.current_ma(), 3),
            "power_mw": round(self.power_mw(), 3),
            "ts_ms": time.ticks_ms(),
        }